experimental/sdk/webpubsub-graphql-subscribe/src/WpsWebSocketServer.ts

```diff
@@ -0,0 +1,135 @@
+
+import { WebPubSubServiceClient } from "@azure/web-pubsub";
+import { WebPubSubEventHandler } from "@azure/web-pubsub-express";
+import { ApolloServer} from "apollo-server-express";
+import { execute, GraphQLSchema, subscribe } from "graphql";
+import { SubscriptionServer } from 'subscriptions-transport-ws';
+import express from 'express';
+import WebSocket from "ws";
+import { WpsPubSub } from "./azure-wps-pubsub";
+import {LOG, log, config} from "./utils"
+
+/**
+  * a virtual WebSocket.Server without real server
+ */
+class VirtualWebSocketServer extends WebSocket.Server{
+	protocol: string = "graphql-ws";
+	readyState: number = WebSocket.OPEN;
+
+	constructor() {
+		 super({port: undefined, noServer: true}); 
+	}
+}
+
+/** 
+  * A VirtualWebSocketServer which is binded to an Azure Web PubSub client connection according to `servecClient` and `connectionId`
+  * It overrides method `send` and send all message to its binded connection in JSON format.
+ */
+class SubWpsWebSocketServer extends VirtualWebSocketServer{
```

Comment:
why not having `send` into `WpsWebSocketServer` directly?

Reply:
`WpsWebSocketServer`  corresponds to [`this.wsServer`](https://github.com/apollographql/subscriptions-transport-ws/blob/9019f95592ad4d9e3cd17e54aa6b3af35dd26e7f/src/server.ts#L132). Then the event [`connection`](https://github.com/apollographql/subscriptions-transport-ws/blob/9019f95592ad4d9e3cd17e54aa6b3af35dd26e7f/src/server.ts#L186), whose callback function is [`connectionHandler`](https://github.com/apollographql/subscriptions-transport-ws/blob/9019f95592ad4d9e3cd17e54aa6b3af35dd26e7f/src/server.ts#L138), will be listened by it.

Input parameter of callback function `connectionHandler`  is `(socket: WebSocket, request: IncomingMessage)`

In this function , `socket` is bind to `request` via `connectionContext`.  
Then a series of [events](https://github.com/apollographql/subscriptions-transport-ws/blob/9019f95592ad4d9e3cd17e54aa6b3af35dd26e7f/src/server.ts#L181) will be registered to this `socket`.

It means that method `send` is actually called by `socket` in `connectionHandler`. 
And `socket` corresponds to `SubWpsWebSocketServer` in my code and we override method `send` in `SubWpsWebSocketServer` rather than in `WpsWebSocketServer `

Reply:
I see, so this is the virtual client connection context

Reply:
yes, so method `send` is overrided in  `SubWpsWebSocketServer` to get the corresponding connection context.